src/libsyntax/ext/expand.rs

```diff
@@ -367,6 +367,14 @@ pub fn expand_item_mac(it: P<ast::Item>,
         _ => fld.cx.span_bug(it.span, "invalid item macro invocation")
     });
 
+    // check here to detect use of feature-gated or invalid attributes on
+    // macro invoations since they will not be detected after macro expansion
+    for attr in attrs.iter() {
+        feature_gate::check_attribute(&attr, &fld.cx.parse_sess.span_diagnostic,
+                                      &fld.cx.parse_sess.codemap(),
+                                      &fld.cx.ecfg.features.unwrap());
+    }
+
```

Comment:
I believe we still need to stability check macros in expression and statement positions (search for `drop(attrs)` in this file).
I would refactor this out into a new function in `expand.rs` (something like `check_attributes(attr: &ast::Attribute, fld: &mut MacroExpander)`) and then inline and remove `feature_gate::check_attribute`.